// polymer-analyzer/lib/polymer/polymer2-mixin-scanner.js
/**
 * @license
 * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ast_value_1 = require("../javascript/ast-value");
const esutil = require("../javascript/esutil");
const jsdoc = require("../javascript/jsdoc");
const polymer_element_mixin_1 = require("./polymer-element-mixin");
const polymer2_config_1 = require("./polymer2-config");
class Polymer2MixinScanner {
    scan(document, visit) {
        return __awaiter(this, void 0, void 0, function* () {
            const visitor = new MixinVisitor(document);
            yield visit(visitor);
            return visitor._mixins;
        });
    }
}
exports.Polymer2MixinScanner = Polymer2MixinScanner;
class MixinVisitor {
    constructor(document) {
        this._mixins = [];
        this._currentMixin = null;
        this._currentMixinNode = null;
        this._currentMixinFunction = null;
        this._document = document;
    }
    enterAssignmentExpression(node, parent) {
        if (parent.type !== 'ExpressionStatement') {
            return;
        }
        const comment = esutil.getAttachedComment(parent) || '';
        const docs = jsdoc.parseJsdoc(comment);
        if (this._hasPolymerMixinDocTag(docs)) {
            const name = ast_value_1.getIdentifierName(node.left);
            const sourceRange = this._document.sourceRangeForNode(node);
            this._currentMixin = new polymer_element_mixin_1.ScannedPolymerElementMixin({
                name, sourceRange,
            });
            this._currentMixinNode = node;
            this._mixins.push(this._currentMixin);
        }
    }
    enterFunctionDeclaration(node, _parent) {
        const comment = esutil.getAttachedComment(node) || '';
        const docs = jsdoc.parseJsdoc(comment);
        if (this._hasPolymerMixinDocTag(docs)) {
            const name = node.id.name;
            const sourceRange = this._document.sourceRangeForNode(node);
            this._currentMixinFunction = node;
            this._currentMixin = new polymer_element_mixin_1.ScannedPolymerElementMixin({
                name, sourceRange,
            });
            this._currentMixinNode = node;
            this._mixins.push(this._currentMixin);
        }
    }
    leaveFunctionDeclaration(node, _parent) {
        if (this._currentMixinNode === node) {
            this._currentMixin = null;
            this._currentMixinNode = null;
            this._currentMixinFunction = null;
        }
    }
    enterVariableDeclaration(node, _parent) {
        const comment = esutil.getAttachedComment(node) || '';
        const docs = jsdoc.parseJsdoc(comment);
        const isMixin = this._hasPolymerMixinDocTag(docs);
        const sourceRange = this._document.sourceRangeForNode(node);
        if (isMixin) {
            this._currentMixin = new polymer_element_mixin_1.ScannedPolymerElementMixin({
                sourceRange,
            });
            this._currentMixinNode = node;
            this._mixins.push(this._currentMixin);
        }
    }
    leaveVariableDeclaration(node, _parent) {
        if (this._currentMixinNode === node) {
            this._currentMixin = null;
            this._currentMixinNode = null;
            this._currentMixinFunction = null;
        }
    }
    enterVariableDeclarator(node, _parent) {
        if (this._currentMixin != null && this._currentMixinFunction == null) {
            this._currentMixin.name = node.id.name;
        }
    }
    enterFunctionExpression(node, _parent) {
        if (this._currentMixin != null && this._currentMixinFunction == null) {
            this._currentMixinFunction = node;
        }
    }
    enterArrowFunctionExpression(node, _parent) {
        if (this._currentMixin != null && this._currentMixinFunction == null) {
            this._currentMixinFunction = node;
        }
    }
    enterClassExpression(node, parent) {
        if (parent.type !== 'ReturnStatement' &&
            parent.type !== 'ArrowFunctionExpression') {
            return;
        }
        this._handleClass(node);
    }
    enterClassDeclaration(node, _parent) {
        const comment = esutil.getAttachedComment(node) || '';
        const docs = jsdoc.parseJsdoc(comment);
        const isMixinClass = this._hasPolymerMixinClassDocTag(docs);
        if (isMixinClass) {
            this._handleClass(node);
        }
    }
    _handleClass(node) {
        const mixin = this._currentMixin;
        if (mixin == null) {
            return;
        }
        const comment = esutil.getAttachedComment(node) || '';
        const docs = jsdoc.parseJsdoc(comment);
        const config = polymer2_config_1.getConfig(node);
        const properties = config && polymer2_config_1.getProperties(config, this._document);
        mixin.description = docs.description ? docs.description.trim() : '';
        mixin.events = esutil.getEventComments(node);
        // mixin.sourceRange = this._document.sourceRangeForNode(node);
        if (properties) {
            properties.forEach((p) => mixin.addProperty(p));
        }
        return mixin;
    }
    _hasPolymerMixinDocTag(docs) {
        const elementTags = docs.tags &&
            docs.tags.filter((t) => t.tag === 'polymerMixin');
        return elementTags && elementTags.length >= 1;
    }
    _hasPolymerMixinClassDocTag(docs) {
        const elementTags = docs.tags &&
            docs.tags.filter((t) => t.tag === 'polymerMixinClass');
        return elementTags && elementTags.length >= 1;
    }
}
